README.md
# EditOnline
A online file editor by Python base on Ace.js.

## How to use
1.run command: python EditOnline.py

2.use broswer open 'http://127.0.0.1:8000'



CHANGES.rst
CHANGES
===============
0.1.0

- test

0.1.2

- first version

0.1.3

- setup bug fix

0.1.4

- 1.complete document
- 2.fix working directory(-d) option bug
- 3.change to use cdnjs static resource

0.1.5

- 1.using **ThreadingHTTPServer** to handle concurrency request
- 2.import **static** resource to avoid the slow speed of cdnjs in China, then the EditOnline size increased >.<
- 3.bind address support, see **-h**
- 4.add EditOnline.py to scripts

0.1.6

- 1.fix libdir bug
- 2.change package info

0.1.7

- 1.wsgi suport


0.1.9

- 1.shell bug

0.1.10

- 1.fix wsgi bug

0.1.11

- 1.short key change
- 2.add toolbar

0.1.12

- 1.short key help

0.1.13

- 1.fix 401 bug

0.1.14

- 1.add fontsize set

main.py
# -*- coding: UTF-8 -*
'''
@author: sintrb
'''
"""Simple HTTP Server.

This module builds on BaseHTTPServer by implementing the standard GET
and HEAD requests in a fairly straightforward manner.

"""


__version__ = "0.1"

import os
import posixpath
import BaseHTTPServer
import urllib
import urlparse
import cgi
import sys
import shutil
import mimetypes
import json
try:
	from cStringIO import StringIO
except ImportError:
	from StringIO import StringIO

libdir = os.path.dirname(__file__)

class EditOnlineRequestHandler(BaseHTTPServer.BaseHTTPRequestHandler):
	server_version = "EditOnline/" + __version__
	def do_GET(self):
		f = self.send_head()
		if f:
			try:
				self.copyfile(f, self.wfile)
			finally:
				f.close()

	def do_POST(self):
		self.path = self.path.replace('..', '')

		length = int(self.headers.getheader('content-length'))
		body = self.rfile.read(length)
		path = self.translate_path(self.path).replace('~editor','')
		if not os.path.exists(os.path.dirname(path)):
			os.makedirs(os.path.dirname(path))
		try:
			with open(path, 'wb') as f:
				f.write(body)
		except:
			self.send_error(404, "No permission to list directory")
		
		f = StringIO()
		f.write(json.dumps({
						'ok':True
						}))
		
		self.send_response(200)
		self.send_header("Content-type", 'text/json')
		self.send_header("Content-Length", str(f.tell()))
		self.end_headers()
		f.seek(0)
		try:
			self.copyfile(f, self.wfile)
		finally:
			f.close()

	def send_head(self):
		self.path = self.path.replace('..', '')
		path = self.translate_path(self.path)
		f = None
		# print self.path
		if self.path.endswith('~editor'):
			try:
				res = ''.join(open(os.path.join(libdir, 'editor.html')).readlines())
				cxt = {
					'path':self.path,
					'version':__version__,
					'exists':'true' if os.path.exists(self.translate_path(self.path.replace('~editor',''))) else 'false',
					}
				
				for k, v in cxt.items():
					res = res.replace('{{%s}}' % k, v)

				f = StringIO()
				f.write(res)
				
				self.send_response(200)
				self.send_header("Content-type", 'text/html')
				self.send_header("Content-Length", str(f.tell()))
				self.end_headers()
				f.seek(0)
				return f
			except IOError:
				self.send_error(404, "File not found")
				return None
		elif os.path.isdir(path):
			parts = urlparse.urlsplit(self.path)
			if not parts.path.endswith('/'):
				# redirect browser - doing basically what apache does
				self.send_response(301)
				new_parts = (parts[0], parts[1], parts[2] + '/',
							 parts[3], parts[4])
				new_url = urlparse.urlunsplit(new_parts)
				self.send_header("Location", new_url)
				self.end_headers()
				return None
			else:
				return self.list_directory(path)
		else:
			path = self.path
			if path.startswith('/eo.static/'):
				# static
				path = os.path.join(libdir, path.replace('/eo.static/', 'static/'))
			else:
				path = self.translate_path(path)
			ctype = self.guess_type(path)
			try:
				# Always read in binary mode. Opening files in text mode may cause
				# newline translations, making the actual size of the content
				# transmitted *less* than the content-length!
				f = open(path, 'rb')
			except IOError:
				self.send_error(404, "File not found")
				return None
			try:
				self.send_response(200)
				self.send_header("Content-type", ctype)
				fs = os.fstat(f.fileno())
				self.send_header("Content-Length", str(fs[6]))
				self.send_header("Last-Modified", self.date_time_string(fs.st_mtime))
				self.end_headers()
				return f
			except:
				f.close()
				raise

	def list_directory(self, path):
		try:
			list = filter(lambda s:not s.startswith('.'), os.listdir(path))
		except os.error:
			self.send_error(404, "No permission to list directory")
			return None
		list.sort(key=lambda a: (' ' if os.path.isdir(a) else '') + a.lower())
		f = StringIO()
		displaypath = cgi.escape(urllib.unquote(self.path))
		f.write('<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">')
		f.write("<html>\n<title>Directory listing for %s</title>\n" % displaypath)
		f.write("<body>\n<h2>Directory listing for %s</h2>\n" % displaypath)
		f.write("<hr>\n<ul>\n")
		for name in list:
			fullname = os.path.join(path, name)
			displayname = linkname = name
			# Append / for directories or @ for symbolic links
			if os.path.isdir(fullname):
				displayname = name + "/"
				linkname = name + "/"
			if os.path.islink(fullname):
				displayname = name + "@"
				# Note: a link to a directory displays with @ and links with /
			isdir = os.path.isdir(fullname)
			f.write('<li><a href="%s"%s>%s</a>\n'
					% (urllib.quote(linkname) + ('' if isdir else '~editor'), '' if isdir else ' target="_blank"', cgi.escape(displayname)))
		f.write("</ul>\n<hr>\n</body>\n</html>\n")
		length = f.tell()
		f.seek(0)
		self.send_response(200)
		encoding = sys.getfilesystemencoding()
		self.send_header("Content-type", "text/html; charset=%s" % encoding)
		self.send_header("Content-Length", str(length))
		self.end_headers()
		return f

	def translate_path(self, path):
		"""Translate a /-separated PATH to the local filename syntax.

		Components that mean special things to the local file system
		(e.g. drive or directory names) are ignored.  (XXX They should
		probably be diagnosed.)

		"""
		# abandon query parameters
		path = path.split('?', 1)[0]
		path = path.split('#', 1)[0]
		# Don't forget explicit trailing slash when normalizing. Issue17324
		trailing_slash = path.rstrip().endswith('/')
		path = posixpath.normpath(urllib.unquote(path))
		words = path.split('/')
		words = filter(None, words)
		path = os.getcwd()
		for word in words:
			drive, word = os.path.splitdrive(word)
			head, word = os.path.split(word)
			if word in (os.curdir, os.pardir): continue
			path = os.path.join(path, word)
		if trailing_slash:
			path += '/'
		return path

	def guess_type(self, path):
		"""Guess the type of a file.

		Argument is a PATH (a filename).

		Return value is a string of the form type/subtype,
		usable for a MIME Content-type header.

		The default implementation looks the file's extension
		up in the table self.extensions_map, using application/octet-stream
		as a default; however it would be permissible (if
		slow) to look inside the data to make a better guess.

		"""

		base, ext = posixpath.splitext(path)
		if ext in self.extensions_map:
			return self.extensions_map[ext]
		ext = ext.lower()
		if ext in self.extensions_map:
			return self.extensions_map[ext]
		else:
			return self.extensions_map['']

	if not mimetypes.inited:
		mimetypes.init()  # try to read system mime.types
	extensions_map = mimetypes.types_map.copy()
	extensions_map.update({
		'': 'application/octet-stream',  # Default
		'.py': 'text/plain',
		'.c': 'text/plain',
		'.h': 'text/plain',
		})

	def copyfile(self, source, outputfile):
		shutil.copyfileobj(source, outputfile)

def test(HandlerClass=EditOnlineRequestHandler, ServerClass=BaseHTTPServer.HTTPServer):
	BaseHTTPServer.test(HandlerClass, ServerClass)

if __name__ == '__main__':
	test()






EditOnline/wsgi.py
"""
WSGI warp
"""

import os
import wsgiref.util


import EditOnline
try:
    from cStringIO import StringIO
except ImportError:
    from StringIO import StringIO

class HeadsWarp(object):
    environ = None
    def __init__(self, environ):
        self.environ = environ
    def getheader(self, name):
        name = name.upper().replace('-', '_')
        return self.environ.get('HTTP_%s' % (name), self.environ.get(name))

class WSGIHandler(EditOnline.EditOnlineRequestHandler):
    path = None
    headers = None
    
    response_code = 200
    response_message = 'OK'
    request_version = 'HTTP/1.0'
    response_headermap = {
                          'Content-Type':'text/html'
                          }
    
    def __init__(self, request, client_address, server):
        self.request = request
        self.client_address = client_address
        self.server = server
        self.setup()
        try:
            self.handle()
        finally:
            self.finish()
    
    def setup(self):
        pass
    
    def finish(self):
        pass
    
    def handle(self):
        pass
    
    def handle_one_request(self):
        pass
    
    def send_response(self, code, message=None):
        self.response_code = code
        self.response_message = message
    
    def send_header(self, keyword, value):
        self.response_headermap[keyword] = value
    
    @property
    def response_headers(self):
        return self.response_headermap.items()

if os.environ.get('WSGI_PARAMS'):
    argv = [v.strip() for v in os.environ.get('WSGI_PARAMS').split(' ') if v.strip()]
    EditOnline.config(argv)

def application(environ, start_response):    
    handler = WSGIHandler(None, (environ['REMOTE_ADDR'], None), None)
#     for k, v in environ.items():
#         print k, v
    handler.path = environ['PATH_INFO'] + ('?' + environ['QUERY_STRING'] if environ['QUERY_STRING'] else '')
    handler.rfile = environ['wsgi.input']
    handler.headers = HeadsWarp(environ)
    handler.wfile = StringIO()
    handler.request_version = environ['SERVER_PROTOCOL']
    handler.command = environ['REQUEST_METHOD']
    method = 'do_%s' % (environ['REQUEST_METHOD'])
    if hasattr(handler, method):
        handler_method = getattr(handler, method)
        handler_method()
    handler.wfile.seek(0)
    res = handler.wfile.read()[2:]
    if res.startswith('\n\n'):
        res = res[2:]
    if res:
        handler.send_header("Content-Length", str(len(res)))
    
    status = '%s %s' % (handler.response_code or 200, handler.response_message or 'OK')
    headers = [(k, v) for k, v in handler.response_headers if not wsgiref.util.is_hop_by_hop(k)]
    
    start_response(status, headers)

    

    return [res, ]


EditOnline/static/config.js
var config = {
	"modes":[
		{key:"abap", name:"ABAP"},
		{key:"abc", name:"ABC"},
		{key:"actionscript", name:"ActionScript"},
		{key:"ada", name:"ADA"},
		{key:"apache_conf", name:"Apache Conf"},
		{key:"asciidoc", name:"AsciiDoc"},
		{key:"assembly_x86", name:"Assembly x86"},
		{key:"autohotkey", name:"AutoHotKey"},
		{key:"batchfile", name:"BatchFile"},
		{key:"c_cpp", name:"C and C++"},
		{key:"c9search", name:"C9Search"},
		{key:"cirru", name:"Cirru"},
		{key:"clojure", name:"Clojure"},
		{key:"cobol", name:"Cobol"},
		{key:"coffee", name:"CoffeeScript"},
		{key:"coldfusion", name:"ColdFusion"},
		{key:"csharp", name:"C#"},
		{key:"css", name:"CSS"},
		{key:"curly", name:"Curly"},
		{key:"d", name:"D"},
		{key:"dart", name:"Dart"},
		{key:"diff", name:"Diff"},
		{key:"dockerfile", name:"Dockerfile"},
		{key:"dot", name:"Dot"},
		{key:"dummy", name:"Dummy"},
		{key:"dummysyntax", name:"DummySyntax"},
		{key:"eiffel", name:"Eiffel"},
		{key:"ejs", name:"EJS"},
		{key:"elixir", name:"Elixir"},
		{key:"elm", name:"Elm"},
		{key:"erlang", name:"Erlang"},
		{key:"forth", name:"Forth"},
		{key:"ftl", name:"FreeMarker"},
		{key:"gcode", name:"Gcode"},
		{key:"gherkin", name:"Gherkin"},
		{key:"gitignore", name:"Gitignore"},
		{key:"glsl", name:"Glsl"},
		{key:"gobstones", name:"Gobstones"},
		{key:"golang", name:"Go"},
		{key:"groovy", name:"Groovy"},
		{key:"haml", name:"HAML"},
		{key:"handlebars", name:"Handlebars"},
		{key:"haskell", name:"Haskell"},
		{key:"haxe", name:"haXe"},
		{key:"html", name:"HTML"},
		{key:"html_elixir", name:"HTML (Elixir)"},
		{key:"html_ruby", name:"HTML (Ruby)"},
		{key:"ini", name:"INI"},
		{key:"io", name:"Io"},
		{key:"jack", name:"Jack"},
		{key:"jade", name:"Jade"},
		{key:"java", name:"Java"},
		{key:"javascript", name:"JavaScript"},
		{key:"json", name:"JSON"},
		{key:"jsoniq", name:"JSONiq"},
		{key:"jsp", name:"JSP"},
		{key:"jsx", name:"JSX"},
		{key:"julia", name:"Julia"},
		{key:"latex", name:"LaTeX"},
		{key:"lean", name:"Lean"},
		{key:"less", name:"LESS"},
		{key:"liquid", name:"Liquid"},
		{key:"lisp", name:"Lisp"},
		{key:"livescript", name:"LiveScript"},
		{key:"logiql", name:"LogiQL"},
		{key:"lsl", name:"LSL"},
		{key:"lua", name:"Lua"},
		{key:"luapage", name:"LuaPage"},
		{key:"lucene", name:"Lucene"},
		{key:"makefile", name:"Makefile"},
		{key:"markdown", name:"Markdown"},
		{key:"mask", name:"Mask"},
		{key:"matlab", name:"MATLAB"},
		{key:"maze", name:"Maze"},
		{key:"mel", name:"MEL"},
		{key:"mushcode", name:"MUSHCode"},
		{key:"mysql", name:"MySQL"},
		{key:"nix", name:"Nix"},
		{key:"nsis", name:"NSIS"},
		{key:"objectivec", name:"Objective-C"},
		{key:"ocaml", name:"OCaml"},
		{key:"pascal", name:"Pascal"},
		{key:"perl", name:"Perl"},
		{key:"pgsql", name:"pgSQL"},
		{key:"php", name:"PHP"},
		{key:"powershell", name:"Powershell"},
		{key:"praat", name:"Praat"},
		{key:"prolog", name:"Prolog"},
		{key:"properties", name:"Properties"},
		{key:"protobuf", name:"Protobuf"},
		{key:"python", name:"Python"},
		{key:"r", name:"R"},
		{key:"razor", name:"Razor"},
		{key:"rdoc", name:"RDoc"},
		{key:"rhtml", name:"RHTML"},
		{key:"rst", name:"RST"},
		{key:"ruby", name:"Ruby"},
		{key:"rust", name:"Rust"},
		{key:"sass", name:"SASS"},
		{key:"scad", name:"SCAD"},
		{key:"scala", name:"Scala"},
		{key:"scheme", name:"Scheme"},
		{key:"scss", name:"SCSS"},
		{key:"sh", name:"SH"},
		{key:"sjs", name:"SJS"},
		{key:"smarty", name:"Smarty"},
		{key:"snippets", name:"snippets"},
		{key:"soy_template", name:"Soy Template"},
		{key:"space", name:"Space"},
		{key:"sql", name:"SQL"},
		{key:"sqlserver", name:"SQLServer"},
		{key:"stylus", name:"Stylus"},
		{key:"svg", name:"SVG"},
		{key:"swift", name:"Swift"},
		{key:"tcl", name:"Tcl"},
		{key:"tex", name:"Tex"},
		{key:"text", name:"Text"},
		{key:"textile", name:"Textile"},
		{key:"toml", name:"Toml"},
		{key:"twig", name:"Twig"},
		{key:"typescript", name:"Typescript"},
		{key:"vala", name:"Vala"},
		{key:"vbscript", name:"VBScript"},
		{key:"velocity", name:"Velocity"},
		{key:"verilog", name:"Verilog"},
		{key:"vhdl", name:"VHDL"},
		{key:"wollok", name:"Wollok"},
		{key:"xml", name:"XML"},
		{key:"xquery", name:"XQuery"},
		{key:"yaml", name:"YAML"},
		{key:"django", name:"Django"},
	],
	"themegroups":[
		{"name":"bright",
		"themes":[
			{key:"chrome",name:"chrome"},
			{key:"clouds",name:"clouds"},
			{key:"crimson_editor",name:"crimson_editor"},
			{key:"dawn",name:"dawn"},
			{key:"dreamweaver",name:"dreamweaver"},
			{key:"eclipse",name:"eclipse"},
			{key:"github",name:"github"},
			{key:"solarized_light",name:"solarized_light"},
			{key:"textmate",name:"textmate"},
			{key:"tomorrow",name:"tomorrow"},
			{key:"xcode",name:"xcode"}
			]
		},
		{"name":"dark",
		"themes":[
			{key:"clouds_midnight",name:"clouds_midnight"},
			{key:"cobalt",name:"cobalt"},
			{key:"idle_fingers",name:"idle_fingers"},
			{key:"kr_theme",name:"kr_theme"},
			{key:"merbivore",name:"merbivore"},
			{key:"merbivore_soft",name:"merbivore_soft"},
			{key:"mono_industrial",name:"mono_industrial"},
			{key:"monokai",name:"monokai"},
			{key:"pastel_on_dark",name:"pastel_on_dark"},
			{key:"solarized_dark",name:"solarized_dark"},
			{key:"terminal",name:"terminal"},
			{key:"tomorrow_night",name:"tomorrow_night"},
			{key:"tomorrow_night_blue",name:"tomorrow_night_blue"},
			{key:"tomorrow_night_bright",name:"tomorrow_night_bright"},
			{key:"tomorrow_night_eighties",name:"tomorrow_night_eighties"},
			{key:"twilight",name:"twilight"},
			{key:"vibrant_ink",name:"vibrant_ink"}
			]
		}
	]
};

EditOnline/EditOnline.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*
'''
@author: sintrb
'''
"""EditOnline Server.

This module refer to SimpleHTTPServer

"""


__version__ = "0.1.14"

import os
import posixpath
import BaseHTTPServer
import SocketServer
import socket
import urllib
import urlparse
import cgi
import sys
import shutil
import mimetypes
import json
try:
	from cStringIO import StringIO
except ImportError:
	from StringIO import StringIO

libdir = os.path.dirname(__file__)
if not libdir:
	libdir = os.getcwd()

options = {
		'workdir':os.getcwd(),
		'bind':'0.0.0.0',
		'port':8000
		}




class EditOnlineRequestHandler(BaseHTTPServer.BaseHTTPRequestHandler):
	server_version = "EditOnline/" + __version__
	protocol_version = "HTTP/1.1"
	editortmpl = ''
	def check_auth(self):
		if not options.get('auth'):
			return True
		au = self.headers.getheader('authorization')
		if au and len(au) > 6 and au.endswith(options.get('auth')):
			return True
		f = StringIO()
		f.write('<center><h2>401 Unauthorized</h2></center>')
		
		self.send_response(401, "Unauthorized")
		self.send_header("Content-Type", "text/html")
		self.send_header("Content-Length", str(f.tell()))
		self.send_header("WWW-Authenticate", 'Basic realm="%s"' % (options.get('realm') or self.server_version))
		self.send_header('Connection', 'close')
		self.end_headers()
		f.seek(0)
		shutil.copyfileobj(f, self.wfile)
		return False
	
	def do_GET(self):
		if not self.check_auth():
			return
		
		f = self.send_head()
		if f:
			try:
				self.copyfile(f, self.wfile)
			finally:
				f.close()

	def do_POST(self):
		if not self.check_auth():
			return
		
		self.path = self.path.replace('..', '')

		length = int(self.headers.getheader('content-length'))
		body = self.rfile.read(length)
		path = self.translate_path(self.path).replace('~editor', '')
		if not os.path.exists(os.path.dirname(path)):
			os.makedirs(os.path.dirname(path))
		try:
			with open(path, 'wb') as f:
				f.write(body)
		except:
			self.send_error(404, "No permission to list directory")
		
		f = StringIO()
		f.write(json.dumps({
						'ok':True
						}))
		
		self.send_response(200)
		self.send_header("Content-type", 'text/json')
		self.send_header("Content-Length", str(f.tell()))
		self.end_headers()
		f.seek(0)
		try:
			self.copyfile(f, self.wfile)
		finally:
			f.close()

	def send_head(self):
		self.path = self.path.replace('..', '')
		path = self.translate_path(self.path)
		f = None
		# print self.path
		if self.path.endswith('~editor'):
			editortmplfile = os.path.join(libdir, 'editor.html')
			if not EditOnlineRequestHandler.editortmpl or True:
				try:	
					EditOnlineRequestHandler.editortmpl = ''.join(open(editortmplfile).readlines())
				except IOError:
					self.send_error(500, "The editor template file(%s) not found. Maybe need reinstall EditOnline" % editortmplfile)
					return None
			cxt = {
				'path':self.path,
				'version':__version__,
				'realm':options.get('realm') or '',
				'exists':'true' if os.path.exists(self.translate_path(self.path.replace('~editor', ''))) else 'false',
				}
			res = EditOnlineRequestHandler.editortmpl
			for k, v in cxt.items():
				res = res.replace('{{%s}}' % k, v)

			f = StringIO()
			f.write(res)
			
			self.send_response(200)
			self.send_header("Content-type", 'text/html')
			self.send_header("Content-Length", str(f.tell()))
			self.send_header('Connection', 'close')
			self.end_headers()
			f.seek(0)
			return f
		elif os.path.isdir(path):
			parts = urlparse.urlsplit(self.path)
			if not parts.path.endswith('/'):
				# redirect browser - doing basically what apache does
				self.send_response(301)
				new_parts = (parts[0], parts[1], parts[2] + '/',
							 parts[3], parts[4])
				new_url = urlparse.urlunsplit(new_parts)
				self.send_header("Location", new_url)
				self.send_header('Connection', 'close')
				self.end_headers()
				return None
			else:
				return self.list_directory(path)
		else:
			path = self.path
			if path.startswith('/eo.static/'):
				# static
				path = os.path.join(libdir, path.replace('/eo.static/', 'static/'))
			else:
				path = self.translate_path(path)
			ctype = self.guess_type(path)
			try:
				# Always read in binary mode. Opening files in text mode may cause
				# newline translations, making the actual size of the content
				# transmitted *less* than the content-length!
				f = open(path, 'rb')
			except IOError:
				self.send_error(404, "File not found")
				return None
			try:
				self.send_response(200)
				self.send_header("Content-type", ctype)
				fs = os.fstat(f.fileno())
				self.send_header("Content-Length", str(fs[6]))
				self.send_header("Last-Modified", self.date_time_string(fs.st_mtime))
				self.send_header('Connection', 'close')
				self.end_headers()
				return f
			except:
				f.close()
				raise

	def list_directory(self, path):
		try:
			list = filter(lambda s:not s.startswith('.'), os.listdir(path))
		except os.error:
			self.send_error(404, "No permission to list directory")
			return None
		list.sort(key=lambda a: (' ' if os.path.isdir(a) else '') + a.lower())
		f = StringIO()
		displaypath = cgi.escape(urllib.unquote(self.path))
		f.write('<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">')
		f.write("<html>\n<title>Directory listing for %s</title>\n" % displaypath)
		f.write("<body>\n<h2>Directory listing for %s</h2>\n" % displaypath)
		f.write("<hr>\n<ul>\n")
		for name in list:
			fullname = os.path.join(path, name)
			displayname = linkname = name
			# Append / for directories or @ for symbolic links
			if os.path.isdir(fullname):
				displayname = name + "/"
				linkname = name + "/"
			if os.path.islink(fullname):
				displayname = name + "@"
				# Note: a link to a directory displays with @ and links with /
			isdir = os.path.isdir(fullname)
			l1 = '<a href="%s">%s</a>' % (urllib.quote(linkname), cgi.escape(displayname));
			l2 = '<a href="%s~editor" target="_blank">edit</a>' % (urllib.quote(linkname));
			if isdir:
				li = '<li>%s</li>' % l1
			else:
				li = '<li>%s [%s]</li>' % (l1, l2)
			f.write(li)
		f.write("</ul>\n<hr>\n</body>\n</html>\n")
		length = f.tell()
		f.seek(0)
		self.send_response(200)
		encoding = sys.getfilesystemencoding()
		self.send_header("Content-type", "text/html; charset=%s" % encoding)
		self.send_header("Content-Length", str(length))
		self.send_header('Connection', 'close')
		self.end_headers()
		return f

	def translate_path(self, path):
		"""Translate a /-separated PATH to the local filename syntax.

		Components that mean special things to the local file system
		(e.g. drive or directory names) are ignored.  (XXX They should
		probably be diagnosed.)

		"""
		# abandon query parameters
		path = path.split('?', 1)[0]
		path = path.split('#', 1)[0]
		# Don't forget explicit trailing slash when normalizing. Issue17324
		trailing_slash = path.rstrip().endswith('/')
		path = posixpath.normpath(urllib.unquote(path))
		words = path.split('/')
		words = filter(None, words)
		path = options.get('workdir')
		for word in words:
			drive, word = os.path.splitdrive(word)
			head, word = os.path.split(word)
			if word in (os.curdir, os.pardir): continue
			path = os.path.join(path, word)
		if trailing_slash:
			path += '/'
		return path

	def guess_type(self, path):
		"""Guess the type of a file.

		Argument is a PATH (a filename).

		Return value is a string of the form type/subtype,
		usable for a MIME Content-type header.

		The default implementation looks the file's extension
		up in the table self.extensions_map, using application/octet-stream
		as a default; however it would be permissible (if
		slow) to look inside the data to make a better guess.

		"""

		base, ext = posixpath.splitext(path)
		if ext in self.extensions_map:
			return self.extensions_map[ext]
		ext = ext.lower()
		if ext in self.extensions_map:
			return self.extensions_map[ext]
		else:
			return self.extensions_map['']

	if not mimetypes.inited:
		mimetypes.init()  # try to read system mime.types
	extensions_map = mimetypes.types_map.copy()
	extensions_map.update({
		'': 'application/octet-stream',  # Default
		'.py': 'text/plain',
		'.c': 'text/plain',
		'.h': 'text/plain',
		})

	def copyfile(self, source, outputfile):
		shutil.copyfileobj(source, outputfile)


class ThreadingHTTPServer(SocketServer.ThreadingTCPServer):
	allow_reuse_address = 1  # Seems to make sense in testing environment
	def server_bind(self):
		"""Override server_bind to store the server name."""
		SocketServer.TCPServer.server_bind(self)
		host, port = self.socket.getsockname()[:2]
		self.server_name = socket.getfqdn(host)
		self.server_port = port

def start():
	port = options['port'] if 'port' in options else 8000
	server_address = (options['bind'], port)
	httpd = ThreadingHTTPServer(server_address, EditOnlineRequestHandler)
	sa = httpd.socket.getsockname()
	print "Root Directory: %s" % options.get('workdir')
	print "Serving HTTP on", sa[0], "port", sa[1], "..."
	httpd.serve_forever()

def config(argv):
	import getopt
	opts, args = getopt.getopt(argv, "u:p:r:hd:")
	for opt, arg in opts:
		if opt == '-u':
			options['username'] = arg
		elif opt == '-p':
			options['password'] = arg
		elif opt == '-r':
			options['realm'] = arg
		elif opt == '-d':
			options['workdir'] = arg
		elif opt == '-h':
			print 'Usage: python -m EditOnline [-u username] [-p password] [-r realm] [-d workdir] [bindaddress:port | port]'
			print 'Report bugs to <sintrb@gmail.com>'
			exit()

	if options.get('username') and options.get('password'):
		import base64
		options['auth'] = base64.b64encode('%s:%s' % (options.get('username'), options.get('password')))
	if len(args) > 0:
		bp = args[0]
		if ':' in bp:
			options['bind'] = bp[0:bp.index(':')]
			options['port'] = int(bp[bp.index(':') + 1:])
		else:
			options['bind'] = '0.0.0.0'
			options['port'] = int(bp)

def main():
	config(sys.argv[1:])
	start()
	
if __name__ == '__main__':
	main()
